packages/main/src/Toolbar.ts

```diff
@@ -0,0 +1,590 @@
+	 */
+	@property({ type: Boolean })
+	resizing!: boolean;
+
+	/**
+	* Calculated width of the whole toolbar.
+	* @private
+	*/
+	@property({ type: Integer })
+	width!: number;
+
+	/**
+	* @private
+	* Calculated width of all the Toolbar items.
+	*/
+	@property({ type: Integer })
+	contentWidth!: number;
+
+	/**
+	 * Notifies the toolbar if it should show the items in a reverse way if Toolbar Popover needs to be placed on "Top" position.
+	 * @private
+	 * @type {sap.ui.webc.main.types.ToolbarAlign}
+	 */
+	@property({ type: Boolean })
+	reverseOverflow!: boolean;
+
+	/**
+ 	* Slotted Toolbar items
+	* @type {sap.ui.webc.main.IToolbarItem[]}
+	* @name sap.ui.webc.main.Toolbar.prototype.default
+	* @slot actions
+	* @public
+	*/
+	@slot({ "default": true, type: HTMLElement, invalidateOnChildChange: true })
+	actions!: Array<ToolbarItem>
+
+	_onResize!: ResizeObserverCallback;
+
+	constructor() {
+		super();
+
+		// actions displayed in the bar
+		this.actionsToBar = [];
+
+		// actions displayed in the toolbar overflow popover
+		this.actionsToOverflow = [];
+
+		// store for actions width
+		this.ACTIONS_WIDTH_MAP = {};
+
+		// indicates the actions have been measured and the layout can be calculated
+		this.actionsWidthMeasured = false;
+
+		// indicates if the bar is being reized at the moment
+		this.resizing = false;
+
+		// resize handler
+		this._onResize = this.onResize.bind(this);
+	}
+
+	static get OVERFLOW_BTN_SIZE(): number {
+		return 44;
+	}
+
+	static get ACTION_MARGIN() {
+		return 4;
+	}
+
+	static get PADDING() {
+		return 16;
+	}
+
+	/**
+	 * Lifecycle methods
+	 */
+	onEnterDOM() {
+		ResizeHandler.register(this, this._onResize);
+	}
+
+	onExitDOM() {
+		ResizeHandler.deregister(this, this._onResize);
+	}
+
+	async onAfterRendering() {
+		if (this.resizing) {
+			this.resizing = false;
+			return;
+		}
+
+		await renderFinished();
+		this.storeActionsWidth();
+		this.doLayout();
+	}
+
+	/**
+	 * Layout management
+	 */
+	doLayout(forceLayout = false) {
+		const containerWidth = this.getContainerWidth();
+		const contentWidth: number = this.getAllActionsWidth();
+		const contentOverflows = contentWidth + Toolbar.OVERFLOW_BTN_SIZE > containerWidth;
+
+		// skip calculation if the width has not been changed
+		if (!forceLayout && this.width === containerWidth) {
+			return;
+		}
+
+		if (contentOverflows) {
+			this.distributeActions(contentWidth - containerWidth);
+		} else {
+			this.displayAllActionsIntoBar();
+		}
+
+		this.width = containerWidth;
+
+		if (this.contentWidth !== contentWidth) {
+			this.contentWidth = contentWidth;
+			this.fireEvent("contentWidthChange", { contentWidth });
+		}
+	}
+
+	storeActionsWidth() {
+		let totalWidth = 0;
+
+		this.movableActions.forEach((action: ToolbarItem) => {
+			const actionWidth = this.getActionWidth(action);
+			const id: string = action._id;
+			totalWidth += actionWidth;
+			this.ACTIONS_WIDTH_MAP[id] = actionWidth;
+		});
+
+		this.ACTIONS_WIDTH_MAP.width = totalWidth;
```

Comment:
I don't mind much, but storing the total width on the map object is a bit strange. Does the TS not complain that the object has no `width` property? Also, the map could be a true `Map` (not generic object). Thus, the TS type will aslo be better. Please at least change this to a true Map, even if you decide to keep the width on it.